File: src/test-dirichlet.cpp
#include <testthat.h>
#include "dirichlet.h"

context("dirichlet") {

  test_that("Dirichlet density can be computed") {
    arma::vec x = arma::zeros<arma::vec>(3);
    x(0) = 0.5; x(1) = 0.3; x(2) = 0.2;
    arma::vec concentration = arma::zeros<arma::vec>(3);
    concentration(0) = 1; concentration(1) = 2; concentration(2) = 3;
    double density = ddirichlet_cpp(x, concentration, false);
    int factor = 1000;
    double density_rounded = round(density * factor) / factor;
    expect_true(density_rounded == 0.720);
    double density_log = ddirichlet_cpp(x, concentration, true);
    double density_log_rounded = round(density_log * factor) / factor;
    expect_true(density_log_rounded == -0.329);
  }

  test_that("Dirichlet can be drawn") {
    arma::vec concentration = arma::zeros<arma::vec>(3);
    concentration(0) = 1; concentration(1) = 2; concentration(2) = 3;
    arma::vec draw = rdirichlet_cpp(concentration);
    expect_true(draw.size() == 3);
    expect_true(abs(sum(draw) - 1) < 0.001);
  }

}

File: tests/testthat/test-dirichlet.R
test_that("Dirichlet density can be computed", {
  x <- c(0.5, 0.3, 0.2)
  concentration <- 1:3
  factor <- 1000
  expect_equal(
    round(ddirichlet(x = x, concentration = concentration, log = FALSE) * factor) / factor,
    0.72
  )
  expect_equal(
    round(ddirichlet(x = x, concentration = concentration, log = TRUE) * factor) / factor,
    -0.329
  )
  expect_error(
    ddirichlet(x = c(0.5, 0.3, 0.3), concentration = concentration),
    "Input `x` is bad"
  )
  expect_equal(ddirichlet(x = c(0.1, 0.2, 0.7), concentration = c(1, 1, 1)), 2)
})

test_that("Dirichlet can be drawn", {
  concentration <- 1:3
  expect_length(rdirichlet(concentration = concentration), 3)
  checkmate::expect_matrix(
    rdirichlet(n = 5, concentration = concentration),
    ncols = 3, nrows = 5
  )
  expect_error(
    rdirichlet(concentration = diag(3)),
    "Input `concentration` is bad"
  )
  checkmate::expect_matrix(
    rdirichlet(n = 5, concentration = 1),
    ncols = 1, nrows = 5
  )
})
